File: Chapter3/ros2_ws/src/avbook_cpp_nodes/src/voxel_grid_filter_node.cpp
#include <functional>
#include <string>

#include <pcl/filters/voxel_grid.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl_conversions/pcl_conversions.h>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>

class VoxelGridFilterNode final : public rclcpp::Node
{
public:
  VoxelGridFilterNode() : rclcpp::Node("voxel_grid_filter_node")
  {
    input_topic_ = this->declare_parameter<std::string>("input_topic", "/points_raw");
    output_topic_ = this->declare_parameter<std::string>("output_topic", "/points_voxel");
    leaf_size_ = this->declare_parameter<double>("leaf_size", 0.2);
    if (leaf_size_ <= 0.0) {
      RCLCPP_WARN(this->get_logger(), "leaf_size <= 0; using 0.2");
      leaf_size_ = 0.2;
    }

    pub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>(output_topic_, rclcpp::QoS(10));
    sub_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
      input_topic_, rclcpp::QoS(10),
      std::bind(&VoxelGridFilterNode::on_cloud, this, std::placeholders::_1));

    RCLCPP_INFO(
      this->get_logger(),
      "VoxelGrid %s -> %s (leaf_size=%.3f m)",
      input_topic_.c_str(),
      output_topic_.c_str(),
      leaf_size_);
  }

private:
  void on_cloud(const sensor_msgs::msg::PointCloud2::ConstSharedPtr msg)
  {
    pcl::PointCloud<pcl::PointXYZ> cloud_in;
    pcl::fromROSMsg(*msg, cloud_in);
    if (cloud_in.empty()) {
      return;
    }

    pcl::VoxelGrid<pcl::PointXYZ> voxel;
    voxel.setInputCloud(cloud_in.makeShared());
    voxel.setLeafSize(
      static_cast<float>(leaf_size_),
      static_cast<float>(leaf_size_),
      static_cast<float>(leaf_size_));

    pcl::PointCloud<pcl::PointXYZ> cloud_out;
    voxel.filter(cloud_out);

    sensor_msgs::msg::PointCloud2 out_msg;
    pcl::toROSMsg(cloud_out, out_msg);
    out_msg.header = msg->header;
    pub_->publish(out_msg);
  }

  std::string input_topic_;
  std::string output_topic_;
  double leaf_size_{0.2};

  rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr sub_;
  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr pub_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<VoxelGridFilterNode>());
  rclcpp::shutdown();
  return 0;
}

File: Chapter3/ros2_ws/src/avbook_py_nodes/avbook_py_nodes/voxel_grid_filter_node.py
import numpy as np
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from sensor_msgs.msg import PointCloud2
from sensor_msgs_py import point_cloud2


class VoxelGridFilterNode(Node):
    def __init__(self) -> None:
        super().__init__("voxel_grid_filter_node")

        self._input_topic = self.declare_parameter("input_topic", "/points_raw").value
        self._output_topic = self.declare_parameter("output_topic", "/points_voxel").value
        self._leaf_size = float(self.declare_parameter("leaf_size", 0.2).value)
        if self._leaf_size <= 0.0:
            self.get_logger().warn("leaf_size <= 0; using 0.2")
            self._leaf_size = 0.2

        self._pub = self.create_publisher(PointCloud2, self._output_topic, 10)
        self._sub = self.create_subscription(PointCloud2, self._input_topic, self._on_cloud, 10)

        self.get_logger().info(
            f"VoxelGrid {self._input_topic} -> {self._output_topic} (leaf_size={self._leaf_size:.3f} m)"
        )

    def _on_cloud(self, msg: PointCloud2) -> None:
        pts_iter = point_cloud2.read_points(msg, field_names=("x", "y", "z"), skip_nans=True)
        pts = np.fromiter(pts_iter, dtype=[("x", np.float32), ("y", np.float32), ("z", np.float32)])
        if pts.size == 0:
            return

        points = np.stack([pts["x"], pts["y"], pts["z"]], axis=1).astype(np.float32, copy=False)

        voxel = np.floor(points / self._leaf_size).astype(np.int32)
        _, inverse = np.unique(voxel, axis=0, return_inverse=True)

        counts = np.bincount(inverse).astype(np.float32)
        sums_x = np.bincount(inverse, weights=points[:, 0])
        sums_y = np.bincount(inverse, weights=points[:, 1])
        sums_z = np.bincount(inverse, weights=points[:, 2])

        centroids = np.stack([sums_x / counts, sums_y / counts, sums_z / counts], axis=1).astype(
            np.float32, copy=False
        )

        out = point_cloud2.create_cloud_xyz32(msg.header, centroids.tolist())
        self._pub.publish(out)


def main() -> None:
    rclpy.init()
    node = VoxelGridFilterNode()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()
